crates/db_connection_pool/src/dbconnection/postgresconn.rs

```diff
@@ -136,14 +143,76 @@ impl<'a>
     ) -> Result<SendableRecordBatchStream> {
-        let rows = self.conn.query(sql, params).await.context(QuerySnafu)?;
-        let rec = rows_to_arrow(rows.as_slice()).context(ConversionSnafu)?;
-        let schema = rec.schema();
-        let recs = vec![rec];
-        Ok(Box::pin(MemoryStream::try_new(recs, schema, None)?))
+        // TODO: We should have a way to detect if params have been passed
+        // if they haven't we should use .copy_out instead, because it should be much faster
+        let streamable = self
+            .conn
+            .query_raw(sql, params.iter().copied()) // use .query_raw to get access to the underlying RowStream
+            .await
+            .context(QuerySnafu)?;
+
+        // chunk the stream into groups of rows based on available system memory
+        let chunk_size = determine_chunk_size();
+        let mut stream = streamable
+            .chunks(chunk_size.try_into()?)
+            .boxed()
+            .map(|rows| {
+                let rows = rows
+                    .into_iter()
+                    .collect::<std::result::Result<Vec<_>, _>>()
+                    .context(QuerySnafu)?;
+                let rec = rows_to_arrow(rows.as_slice()).context(ConversionSnafu)?;
+                Ok::<_, PostgresError>(rec)
+            });
+
+        let Some(first_chunk) = stream.next().await else {
+            return Ok(Box::pin(RecordBatchStreamAdapter::new(
+                Arc::new(Schema::empty()),
+                stream::empty(),
+            )));
+        };
+
+        let first_chunk = first_chunk?;
+        let schema = first_chunk.schema(); // like clickhouse, pull out the schema from the first chunk to use in the DataFusion Stream Adapter
+
+        let output_stream = stream! {
+           yield Ok(first_chunk);
+           while let Some(batch) = stream.next().await {
+                match batch {
+                    Ok(batch) => {
+                        yield Ok(batch); // we can yield the batch as-is because we've already converted to Arrow in the chunk map
+                    }
+                    Err(e) => {
+                        println!("Yielding an error");
+                        yield Err(DataFusionError::Execution(format!("Failed to fetch batch: {e}")));
+                    }
+                }
+           }
+        };
+
+        Ok(Box::pin(RecordBatchStreamAdapter::new(
+            schema,
+            output_stream,
+        )))
     }
 
     async fn execute(&self, sql: &str, params: &[&'a (dyn ToSql + Sync)]) -> Result<u64> {
         Ok(self.conn.execute(sql, params).await?)
     }
 }
+
+const BYTES_PER_KILOBYTE: u64 = 1_024;
+const BYTES_PER_MEGABYTE: u64 = 1_024 * BYTES_PER_KILOBYTE;
+const ROWS_PER_MB: u64 = 1; // 1024 rows per GB
+
+fn determine_chunk_size() -> u64 {
+    let mut sys = System::new_all();
+    sys.refresh_all();
+
+    let mem = sys.available_memory() / BYTES_PER_MEGABYTE; // available memory in MB
```

Comment:
This is a relatively expensive operation: https://docs.rs/sysinfo/latest/sysinfo/#good-practice--performance-tips

We only want to fetch the data we actually use.